File: FindPattern.pl
use strict;
use warnings;
use List::Util qw(shuffle);

# Generate a list of random numbers
sub generate_random_list {
    my ($n) = @_;
    my @random_nums;
    for (1..$n) {
        push @random_nums, int(rand(10));
    }
    return \@random_nums;
}

# Find repeating patterns in the list
sub find_time_loops {
    my ($nums, $len) = @_;
    my %pattern_frequency;

    for my $i (0..$#$nums - $len + 1) {
        my $pattern = join(',', @$nums[$i..$i + $len - 1]) . ',';
        $pattern_frequency{$pattern}++;
    }

    # Keep only patterns that occur more than once
    foreach my $key (keys %pattern_frequency) {
        delete $pattern_frequency{$key} if $pattern_frequency{$key} < 2;
    }

    return \%pattern_frequency;
}

# Look for outliers in the repeating patterns
sub look_for_outliers {
    my ($time_loops) = @_;
    my $avg = 0.0;
    my $max = 0;
    my $max_key = '';

    print "\nGefundene Zeitschleifen in timeLoops:\n";
    if (!%$time_loops) {
        print "Keine Zeitschleifen gefunden.\n";
    } else {
        foreach my $key (keys %$time_loops) {
            print "Muster: $key | Wiederholungen: $time_loops->{$key}\n";
            $avg += $time_loops->{$key};
            if ($time_loops->{$key} > $max) {
                $max = $time_loops->{$key};
                $max_key = $key;
            }
        }
    }

    $avg = $avg / (keys %$time_loops) if (keys %$time_loops);

    if ($max - $avg > 3) {
        print "\nVermeintliche Zeitschleife gefunden!\n";
        print "$max_key wurde $max mal wiederholt.\n";
    } else {
        print "\nKeine vermeintliche Zeitschleife gefunden!\n";
    }
}

# Main execution
my $random_nums = generate_random_list(1000);
my $len = 5;
my $random_time_loops = find_time_loops($random_nums, $len);
look_for_outliers($random_time_loops);

File: FindPattern.pas
program TimeLoopFinder;

uses
  SysUtils, Classes;

type
  TStringMap = record
    Keys: array of string;
    Values: array of Integer;
  end;

procedure AddToMap(var map: TStringMap; key: string; value: Integer);
var
  i: Integer;
begin
  for i := 0 to High(map.Keys) do
  begin
    if map.Keys[i] = key then
    begin
      map.Values[i] := map.Values[i] + value;
      Exit;
    end;
  end;
  SetLength(map.Keys, Length(map.Keys) + 1);
  SetLength(map.Values, Length(map.Values) + 1);
  map.Keys[High(map.Keys)] := key;
  map.Values[High(map.Values)] := value;
end;

procedure RemoveIfLessThanTwo(var map: TStringMap);
var
  i, count: Integer;
begin
  count := 0;
  for i := 0 to High(map.Keys) do
  begin
    if map.Values[i] >= 2 then
    begin
      map.Keys[count] := map.Keys[i];
      map.Values[count] := map.Values[i];
      Inc(count);
    end;
  end;
  SetLength(map.Keys, count);
  SetLength(map.Values, count);
end;

function GenerateRandomList(n: Integer): TList;
var
  i: Integer;
  intPtr: ^Integer;
  randomList: TList;
begin
  Randomize;
  randomList := TList.Create;
  for i := 0 to n - 1 do
  begin
    New(intPtr);
    intPtr^ := Random(10);
    randomList.Add(intPtr);
  end;
  GenerateRandomList := randomList;
end;

function FindTimeLoops(nums: TList; len: Integer): TStringMap;
var
  i, j: Integer;
  pattern: string;
  timeLoops: TStringMap;
begin
  SetLength(timeLoops.Keys, 0);
  SetLength(timeLoops.Values, 0);

  for i := 0 to nums.Count - len do
  begin
    pattern := '';
    for j := i to i + len - 1 do
      pattern := pattern + IntToStr(Integer(PInteger(nums[j])^)) + ',';

    AddToMap(timeLoops, pattern, 1);
  end;

  RemoveIfLessThanTwo(timeLoops);
  FindTimeLoops := timeLoops;
end;

procedure LookForOutliers(timeLoops: TStringMap);
var
  avg: Double;
  max, i: Integer;
  maxKey: string;
begin
  avg := 0.0;
  max := 0;
  maxKey := '';

  WriteLn;
  WriteLn('Gefundene Zeitschleifen in timeLoops:');
  if Length(timeLoops.Keys) = 0 then
    WriteLn('Keine Zeitschleifen gefunden.')
  else
  begin
    for i := 0 to High(timeLoops.Keys) do
    begin
      WriteLn('Muster: ', timeLoops.Keys[i], ' | Wiederholungen: ', timeLoops.Values[i]);
      avg := avg + timeLoops.Values[i];
      if timeLoops.Values[i] > max then
      begin
        max := timeLoops.Values[i];
        maxKey := timeLoops.Keys[i];
      end;
    end;
    avg := avg / Length(timeLoops.Keys);
    if max - avg > 3 then
    begin
      WriteLn;
      WriteLn('Vermeintliche Zeitschleife gefunden!');
      WriteLn(maxKey, ' wurde ', max, ' mal wiederholt.');
    end
    else
    begin
      WriteLn;
      WriteLn('Keine vermeintliche Zeitschleife gefunden!');
    end;
  end;
end;

procedure FreeRandomList(nums: TList);
var
  i: Integer;
begin
  for i := 0 to nums.Count - 1 do
    Dispose(PInteger(nums[i]));
  nums.Free;
end;

var
  randomNums: TList;
  len: Integer;
  randomTimeLoops: TStringMap;

begin
  // 1. Zufällige nums generieren
  randomNums := GenerateRandomList(1000);

  // 2. Zeitschleifen erkennen
  len := 5;
  randomTimeLoops := FindTimeLoops(randomNums, len);

  // 3. Ergebnisse anzeigen
  LookForOutliers(randomTimeLoops);

  // 4. Speicher freigeben
  FreeRandomList(randomNums);
end.
